### PR TITLE
[SPARK-44146][CONNECT] Isolate Spark Connect Session jars and classfiles

### PR DESCRIPTION
**Pull request by vicennial**
### What changes were proposed in this pull request?

This PR follows up on https://github.com/apache/spark/pull/41625 to utilize the classloader/resource isolation in Spark to support multi-user Spark Connect sessions which are isolated from each other (currently, classfiles and jars) and thus, enables multi-user REPLs and UDFs.

- Instead of a single instance of `SparkArtifactManager` handling all the artifact movement, each instance is now responsible for a single `sessionHolder` (i.e a Spark Connect session) which it requires in it's constructor.
- Previously, all artifacts were stored under a common directory `sparkConnectArtifactDirectory` which was initialised in `SparkContext`. Moving forward, all artifacts are instead separated based on the underlying `SparkSession` (using it's `sessionUUID`) they belong to in the format of `ROOT_ARTIFACT_DIR/<sessionUUID>/jars/...`.
- The `SparkConnectArtifactManager` also builds a `JobArtifactSet` [here](https://github.com/apache/spark/pull/41701/files#diff-f833145e80f2b42f54f446a0f173e60e3f5ad657a6ad1f2135bc5c20bcddc90cR157-R168) which is eventually propagated to the executors where the classloader isolation mechanism uses the `uuid` parameter.
- Currently, classfile and jars are isolated but files and archives aren't. 

### Why are the changes needed?

To enable support for multi-user sessions coexisting on a singular Spark cluster. For example, multi-user Scala REPLs/UDFs will be supported with this PR.

### Does this PR introduce _any_ user-facing change?

Yes, multiple Spark Connect REPLs may use a single Spark cluster at once and execute their own UDFs without intefering with each other.

### How was this patch tested?

New unit tests in `ArtifactManagerSuite` + existing tests.

### CODE REVIEW

**Comment by vicennial** [comment]
The PR is ready to review (the falling tests are flaky, I've submitted a rerun request).

**Comment by vicennial** [comment]
@HyukjinKwon @hvanhovell Could you have a look? Thanks!

**Review by HyukjinKwon — APPROVED** [review]
Looks fine from a cursory look but I think we should have @hvanhovell signoff

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala
Comment (hvanhovell) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:92]: We could put this in the session holder right?

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/artifact/SparkConnectArtifactManager.scala
Comment (hvanhovell) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/artifact/SparkConnectArtifactManager.scala:277]: Can we use `addDirectory` instead? The if-absent bit if pretty well protected by this object.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SessionHolder.scala
Comment (hvanhovell) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SessionHolder.scala:110]: This name is a bit too vague for my liking. How about `withContextClassLoader`?

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SessionHolder.scala
Comment (hvanhovell) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SessionHolder.scala:127]: qq, we don't really need to unset this right?

File: connector/connect/server/src/test/scala/org/apache/spark/sql/connect/artifact/ArtifactManagerSuite.scala
Comment (hvanhovell) [inline connector/connect/server/src/test/scala/org/apache/spark/sql/connect/artifact/ArtifactManagerSuite.scala:241]: You could add another session here where you can load Hello. In that cases the classes for the different sessions should not be equal.

**Review by hvanhovell — APPROVED** [review]
LGTM

**Comment by hvanhovell** [comment]
I am merging this to unblock a couple of follow-ups. Please address my comments in a small follow-up.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SessionHolder.scala
Comment (HyukjinKwon) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SessionHolder.scala:127]: yeah I planned to remove this and @vicennial removed it in https://github.com/apache/spark/pull/41789#discussion_r1247634835. This was just a hack I added to avoid additional refactoring.

**Pull request closed without merging**
